frontend/src/lib/components/EditableText.svelte

```diff
@@ -0,0 +1,157 @@
+	import { LoadingIcon } from '$lib/icons';
+	import { ZodString, z } from 'zod';
+	import { _changeProjectName } from '../../routes/project/[project_code]/+page';
+	import IconButton from './IconButton.svelte';
+	import { Form, lexSuperForm, lexSuperValidate } from '$lib/forms';
+
+	export let value: string | undefined | null = undefined;
+	export let disabled = false;
+	export let saveHandler: (newValue: string) => Promise<unknown>;
+	export let placeholder: string | undefined = undefined;
+	export let multiline = false;
+	export let validation: ZodString | undefined = undefined;
+
+	let initialValue: string | undefined | null;
+	let editing = false;
+	let saving = false;
+	const fieldId = 'id' + new Date().getTime();
+
+	let inputElem: HTMLInputElement | HTMLTextAreaElement;
+
+	const formSchema = z.object(validation ? { [fieldId]: validation } : {});
+	let { form, errors, valid, update, reset } = lexSuperForm(
+		formSchema,
+		{ taintedMessage: false },
+	);
+	$: error = $errors[fieldId]?.join(', ');
+
+	function startEditing() {
+		if (disabled) {
+			return;
+		}
+
+		initialValue = value;
+		form.set({[fieldId]: value ?? ''});
+		editing = true;
+	}
+
+	async function save() {
+		const newValue = inputElem.value;
+		if (newValue === initialValue) {
+			editing = false;
+			return;
+		}
+
+		saving = true;
+		editing = false;
+		await saveHandler(newValue);
+		value = newValue;
+		saving = false;
+		
+	}
+
+	function cancel() {
+		editing = false;
+		reset();
+	}
+
+	async function onKeydown(event: KeyboardEvent) {
+		switch (event.key) {
+			case 'Enter':
+				if (multiline && event.ctrlKey) {
+					event.preventDefault();
+					await submit();
+				}
+				break;
+			case 'Esc': // IE/Edge specific value
+			case 'Escape':
+				cancel();
+				break;
+		}
+	}
+
+	async function submit() {
+		await lexSuperValidate($form, formSchema, update);
+		if (!$valid) return;
+		save();
+	}
+</script>
+
+<span class="inline-flex items-end not-prose space-x-2 relative" class:w-full={multiline}>
+	{#if editing || saving}
+		<!-- svelte-ignore a11y-autofocus -->
+		<span class:grow={multiline}>
+			<Form on:submit={submit}>
+				{#if multiline}
+					<textarea
+						id={fieldId}
+						on:keydown={onKeydown}
+						class:textarea-error={error}
+						autofocus
+						bind:value={$form[fieldId]}
+						bind:this={inputElem}
+						readonly={saving}
+						class="textarea textarea-bordered mt-1"
+					/>
+				{:else}
+					<input
+						id={fieldId}
+						on:keydown={onKeydown}
+						class:input-error={error}
+						autofocus
+						bind:value={$form[fieldId]}
+						bind:this={inputElem}
+						readonly={saving}
+						class="input input-bordered mt-1 mb-0"
+					/>
+				{/if}
+			</Form>
+		</span>
+		
+		<IconButton on:click={submit} disabled={saving} icon="i-mdi-check-bold" />
+		<IconButton on:click={cancel} disabled={saving} icon="i-mdi-close-thick" />
+
+		{#if error}
+			<label for={fieldId} class="label absolute -bottom-5 p-0">
```

Comment:
I think there's a typo in `-bottom-5` should that be pb-5?

Reply:
We decided we'll use a tooltip instead